cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -210,43 +197,93 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
+          return Status::NotImplemented("substrait::ReadRel::LocalFiles item (",
+                                        item_uri.ToString(),
+                                        ") with non-filesystem scheme (file:///)");
+        }
+
+        if (item_uri.port() != -1) {
+          return Status::NotImplemented("substrait::ReadRel::LocalFiles item (",
+                                        item_uri.ToString(),
+                                        ") should not have a port number in path");
+        }
+
+        if (!item_uri.query_string().empty()) {
+          return Status::NotImplemented("substrait::ReadRel::LocalFiles item (",
+                                        item_uri.ToString(),
+                                        ") should not have a query string in path");
+        }
+
+        switch (item.file_format_case()) {
+          case substrait::ReadRel::LocalFiles::FileOrFiles::kParquet:
+            format = std::make_shared<dataset::ParquetFileFormat>();
+            break;
+          case substrait::ReadRel::LocalFiles::FileOrFiles::kArrow:
+            format = std::make_shared<dataset::IpcFileFormat>();
+            break;
+          default:
+            // TODO: maybe check for ".feather" or ".arrows"?
+            return Status::NotImplemented(
+                "unsupported file format ",
+                "(see substrait::ReadRel::LocalFiles::FileOrFiles::file_format)");
+        }
+
+        // Handle the URI as appropriate
+        switch (item.path_type_case()) {
           case substrait::ReadRel::LocalFiles::FileOrFiles::kUriFile: {
-            files.emplace_back(path, fs::FileType::File);
+            files.emplace_back(item_uri.path(), fs::FileType::File);
             break;
           }
+
           case substrait::ReadRel::LocalFiles::FileOrFiles::kUriFolder: {
-            fs::FileSelector selector;
-            selector.base_dir = path;
-            selector.recursive = true;
-            ARROW_ASSIGN_OR_RAISE(auto discovered_files,
-                                  filesystem->GetFileInfo(selector));
-            std::move(discovered_files.begin(), discovered_files.end(),
-                      std::back_inserter(files));
+            RETURN_NOT_OK(DiscoverFilesFromDir(filesystem, item_uri.path(), files));
+            break;
+          }
+
+          case substrait::ReadRel::LocalFiles::FileOrFiles::kUriPath: {
+            // Let the filesystem API decide for us if the URI is a file or a directory
+            ARROW_ASSIGN_OR_RAISE(auto file_info,
+                                  filesystem->GetFileInfo(item_uri.path()));
+
+            // push the FileInfo if it's a file; else, recurse into the directory
+            if (file_info.type() == fs::FileType::File) {
+              files.push_back(std::move(file_info));
+            } else if (file_info.type() == fs::FileType::Directory) {
+              RETURN_NOT_OK(DiscoverFilesFromDir(filesystem, item_uri.path(), files));
+            }
```

Comment:
Should probably also handle the `FileInfo::Unknown` and `FileInfo::NotFound` cases (by erroring out perhaps?).

`FileInfo::Unknown` can be got for non-regular files such as pipes, sockets etc.
`FileInfo::NotFound` you probably understand :-)